From a4a741770c57749a289eb57c4f18470630c20eea Mon Sep 17 00:00:00 2001
From: MengLi <meng.li@windriver.com>
Date: Mon, 29 Oct 2018 14:22:02 +0800
Subject: [PATCH] arch: arm: psci: implement psci_system_off interface for
 stratix10

When execute "poweroff" command in Linux, smc instruction is
invoked and then enter el3 mode to execute system off processing.
But in current code, there is only a weak define of psci_system_off,
and vendor doesn't implement this interface to power off system.
So, only make a workaround so that straxi10 board is able to run
command "poweroff" smoothly in linux.

Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm/mach-socfpga/psci.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/arch/arm/mach-socfpga/psci.c b/arch/arm/mach-socfpga/psci.c
index 645ef83..b49c43a 100644
--- a/arch/arm/mach-socfpga/psci.c
+++ b/arch/arm/mach-socfpga/psci.c
@@ -58,3 +58,13 @@ int __secure psci_cpu_on_64(u32 function_id, u64 cpuid, u64 entry_point)
 
 	return ARM_PSCI_RET_SUCCESS;
 }
+
+void __noreturn __secure psci_system_off(void)
+{
+	puts("poweroff ...\n");
+	udelay(50000); /* wait 50 ms */
+	disable_interrupts();
+
+	while (1)
+		;
+}
-- 
2.7.4

